File: built-in plugins dsp/mb-restore-phase-correct.cpp
/**
 * MB Phase Correction
 * Category : effect
 * Type     : eq
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Automatic multi-mic phase alignment and correction
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_RESTORE_PHASE_CORRECT_H
#define MB_RESTORE_PHASE_CORRECT_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbRestorePhaseCorrect : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-restore-phase-correct";
    static constexpr const char* PLUGIN_NAME    = "MB Phase Correction";
    static constexpr const char* PLUGIN_TYPE    = "eq";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float maxDelay = 10f;  // range [1, 50]
    float sensitivity = 0.7f;  // range [0, 1]
    float flipPhase = 0f;  // range [0, 1]
    };

    MbRestorePhaseCorrect() = default;
    ~MbRestorePhaseCorrect() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.maxDelay = std::clamp(params.maxDelay, 1f, 50f);
        params.sensitivity = std::clamp(params.sensitivity, 0f, 1f);
        params.flipPhase = std::clamp(params.flipPhase, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Phase Correction
        return input;
    }
};

#endif // MB_RESTORE_PHASE_CORRECT_H

File: built-in plugins dsp/mb-restore-phase-correct.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbRestorePhaseCorrectPlugin: PluginDefinition = { id: 'mb-restore-phase-correct', slug: 'mb-restore-phase-correct', name: 'MB Phase Correction', category: 'effect', type: 'eq' as any, version: '1.0.0', description: 'Automatic multi-mic phase alignment and correction', author: 'Max Booster', grade: 'A', parameters: [{ id: 'maxDelay', name: 'Max Delay', type: 'float', defaultValue: 10, minValue: 1, maxValue: 50, automatable: false }, { id: 'sensitivity', name: 'Sensitivity', type: 'float', defaultValue: 0.7, minValue: 0, maxValue: 1, automatable: true }, { id: 'flipPhase', name: 'Flip Phase', type: 'float', defaultValue: 0, minValue: 0, maxValue: 1, automatable: false }], defaultPreset: { maxDelay: 10, sensitivity: 0.7, flipPhase: 0 } };

export default MbRestorePhaseCorrectPlugin;
